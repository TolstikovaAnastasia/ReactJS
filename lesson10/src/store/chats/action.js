import { onValue, set } from 'firebase/database';
import { chatsRef, getChatsRefById, getMsgsRefById } from '../../service/firebase';

export const SET_CHAT = 'CHAT::SET_CHAT';
export const ADD_CHAT = 'CHAT::ADD_CHAT';
export const DELETE_CHAT = 'CHAT::DELETE_CHAT';

export const addChat = (newChat) => ({
    type: ADD_CHAT,
    payload: newChat,
});

export const deleteChat = (chatId) => ({
    type: DELETE_CHAT,
    payload: chatId,
});

const setChat = (chats) => ({
    type: SET_CHAT,
    payload: chats,
});

export const initChatsTracking = () => (dispatch) => {
    onValue(chatsRef, (snapshot) => {
        const newChats = [];
        snapshot?.forEach((chat) => {
            newChats.push(chat.val());
        });
        dispatch(setChat(newChats));
    });
};

export const addChatsWithFb = (newChatName) => () => {
    const newId = `chat${Date.now()}`;
    const newChat = {
        id: newId,
        name: newChatName,
    };
    set(getChatsRefById(newId), newChat);
    set(getMsgsRefById(newId), { empty: true });
};

export const deleteChatWithFb = (id) => () => {
    set(getChatsRefById(id), null);
    set(getMsgsRefById(id), null);
};