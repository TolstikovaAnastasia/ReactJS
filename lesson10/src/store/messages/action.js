import { USERS } from '../../utils/constant';
import { onValue, push } from 'firebase/database';
import { getMsgsListRefById, msgsRef } from '../../service/firebase';

export const SET_MESSAGE = 'MESSAGE::SET_MESSAGE';
export const ADD_MESSAGE = 'MESSAGE::ADD_MESSAGE';
export const DELETE_MESSAGE = 'MESSAGE::DELETE_MESSAGE';

export const addMessage = (newMessage, chatId) => ({
    type: ADD_MESSAGE,
    payload: {
        message: newMessage,
        chatId,
    },
});

export const deleteMessage = (messageId, chatId) => ({
    type: DELETE_MESSAGE,
    payload: {
        messageId,
        chatId,
    },
});

export const setMsgs = (msgs) => ({
    type: SET_MESSAGE,
    payload: msgs,
});

let timerId;

export const addMessageWithReply = (newMessage, chatId) => (dispatch) => {
    dispatch(addMessage(newMessage, chatId));

    clearTimeout(timerId);
    if(newMessage.author === USERS.YOU) {
        timerId = setTimeout(() => {
            dispatch(addMessage(
                {
                    text: 'Ваше сообщение могут видеть все участники чата',
                    author: USERS.BOT,
                    id: `msg-${Date.now()}`,
                },
                chatId)
            );
        }, 2500);
    }
};

export const initMsgsTracking = () => (dispatch) => {
    onValue(msgsRef, (snapshot) => {
        const newMsgs = {};
        snapshot?.forEach((chatsMsgsSnap) => {
            newMsgs[chatsMsgsSnap.key] = Object.values(
                chatsMsgsSnap.val().messageList || {}
            );
        });
        dispatch(setMsgs(newMsgs));
    });
};

export const addMsgWithFb = (newMsg, chatId) => () => {
    push(getMsgsListRefById(chatId), newMsg);
};