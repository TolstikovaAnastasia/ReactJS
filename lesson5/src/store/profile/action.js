export const PROFILE_CONTENT = 'PROFILE_CONTENT';