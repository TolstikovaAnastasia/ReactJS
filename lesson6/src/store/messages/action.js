export const ADD_MESSAGE = 'MESSAGE::ADD_MESSAGE';
export const DELETE_MESSAGE = 'MESSAGE::DELETE_MESSAGE';

export const addMessage = (newMessage, chatId) => ({
    type: ADD_MESSAGE,
    payload: {
        message: newMessage,
        chatId,
    }
});

export const deleteMessage = (messageId, chatId) => ({
    type: DELETE_MESSAGE,
    payload: {
        messageId,
        chatId,
    }
});