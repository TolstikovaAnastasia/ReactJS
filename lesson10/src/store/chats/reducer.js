import { ADD_CHAT, DELETE_CHAT, SET_CHAT } from "./action";

const initialState = [];

export const chatsReducer = (state = initialState, action) => {
    switch(action?.type) {
        case ADD_CHAT:
            return [
                ...state,
                action.payload
            ];
        case DELETE_CHAT:
            return state.filter((chat) => chat.id !== action.payload);
        case SET_CHAT:
            return action.payload;
        default:
            return state;
    }
};