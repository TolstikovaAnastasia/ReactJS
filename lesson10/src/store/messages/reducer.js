import { ADD_CHAT, DELETE_CHAT } from "../chats/action";
import { ADD_MESSAGE, SET_MESSAGE } from "./action";

const initialState = {};

export const messagesReducer = (state = initialState, action) => {
    switch(action?.type) {
        case ADD_MESSAGE:
            return {
                ...state,
                [action.payload.chatId]: [
                    ...state[action.payload.chatId],
                    action.payload.message,
                ],
            };
        case ADD_CHAT:
            return {
                ...state,
                [action.payload.id]: [],
            };
        case DELETE_CHAT: {
            const newMessages = { ...state };
            delete newMessages[action.payload];
            return newMessages;
        }
        case SET_MESSAGE: {
            return action.payload;
        }
        default:
            return state;
    }
};