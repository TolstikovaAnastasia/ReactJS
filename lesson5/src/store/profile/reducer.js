import { PROFILE_CONTENT } from "./action";

const initialState = {
    content: false,
};

export const profileReducer = (state = initialState, action) => {
    switch (action?.type) {
        case PROFILE_CONTENT:
            return {
                ...state,
                content: !state.content,
            };
        default:
            return state;
    }
};