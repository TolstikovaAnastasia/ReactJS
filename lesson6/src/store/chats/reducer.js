import { ADD_CHAT, DELETE_CHAT } from "./action";

const initialState = [];

export const chatsReducer = (state = initialState, action) => {
    switch(action?.type) {
        case ADD_CHAT:
            return [
                ...state,
                action.payload
            ];
        case DELETE_CHAT:
            return state.chatList.filter((chat) => chat.Id !== action.payload);
        default:
            return state;
    }
};