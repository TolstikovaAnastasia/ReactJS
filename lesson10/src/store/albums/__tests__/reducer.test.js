import { GET_ALBUMS_REQUEST } from "../action";
import { REQUEST_STATUS } from '../../../utils/constant';
import { albumsReducer } from "../reducer";

describe('albumsReducer tests', () => {
    it('should handle GET_ALBUMS_REQUEST', () => {
        const expected = albumsReducer([], {
            type: GET_ALBUMS_REQUEST,
            status: REQUEST_STATUS.PENDING,
        });
        expect(expected).toMatchSnapshot();
    });
});