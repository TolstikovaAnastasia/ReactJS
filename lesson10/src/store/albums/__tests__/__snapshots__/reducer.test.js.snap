// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`albumsReducer tests should handle GET_ALBUMS_REQUEST 1`] = `
Object {
  "request": Object {
    "error": "",
    "status": 1,
  },
}
`;
